[14.0][WIP][FIX] fix delivery date

mmequignon opened the pull request:
The last 5 commits are ported from https://github.com/OCA/sale-workflow/pull/1968 (postpone delivery dates for backorders, or when ordered qty is modified)

mmequignon committed:
[TMP] fix to resolve calendar conflict with DDMRP

There's a conflict with DDMRP regarding the usage of the
stock.warehouse's calendar_id.
This PR implements a new `calendar2_id` field (already there in v13, see
this commit https://github.com/OCA/sale-workflow/pull/1696/commits/5b0b93a126706b3f32e789b6e751099001b16bcc)
in order to temporary solve this conflict.

sebalix committed:
[FIX] sale_order_line_date: fix commitment_date onchange

In 14.0 there are two methods named `_onchange_commitment_date':

1) one which set the `commitment_date` with the `expected_date` value
   (triggered by `expected_date`)
2) the second which triggers a warning (triggered by `commitment_date`
   only)

The first one is never executed because the second has the precedence,
and it is a good thing because we do not want to automatically set the
`commitment_date` by default!
Odoo has fixed this code in 15.0 by removing the first method and added
the `expected_date` field as trigger on the second one
(see 4031e86f2564378d18f7a930ad45f178c71b2939).

This commit is adding `expected_date` as well on this overridden method
so we are in line with the expected behavior.

sebalix committed:
[FIX] sale_delivery_date: add dependency

Add `sale_order_line_date` as dependency because this addon is
overriding the same methods (`_onchange_commitment_date` and
`_prepare_procurement_values`) and we want to ensure than
`sale_delivery_date` takes the precedence over `sale_order_line_date`.

Also, the other feature of `sale_order_line_date` which sync the
`commitment_date` between SO and its lines is welcome in
`sale_delivery_date`.

mmequignon committed:
[IMP] sale_delivery_date: refactoring

mmequignon committed:
[IMP] sale_delivery_date: test individual methods

mmequignon committed:
[FIX] sale_delivery_date: fix tests

mmequignon committed:
[IMP] sale_delivery_date: convert methods to api.model

mmequignon committed:
[IMP] sale_delivery_date: Recompute backorders dates

mmequignon committed:
[IMP] sale_delivery_date: Postpone dates if sale line qty is modified

jbaudoux commented on sale_delivery_date/models/stock_picking.py:
```diff
@@ -148,3 +148,29 @@ def _onchange_scheduled_date(self):
+        return cutoff
+
+    def _create_backorder(self):
+        res = super()._create_backorder()
+        now = fields.Datetime.now()
+        for picking in res:
+            # If the scheduled_date is before the current datetime, then date_deadline
+            # cannot be satisfied. Therefore, we need to recompute move's dates
+            if picking.scheduled_date < now:
+                for line in picking.move_lines:
+                    dates = line._get_delivery_dates(from_date=now)
+                    line.write(dates)
+        return res
```
This is wrong. You cannot reschedule at backorder creation.
A picking is scheduled based customer/transporter/marketing constrains and if you create a backorder at operational level, the scheduled date has no reason to change.

The use case of rescheduling the open deliveries that could not be processed on the planned day must be tackled differently. For instance, at release channel level.

Can you move this commit out of this PR?

jbaudoux replied:
This relates to my remark here that is still valid https://github.com/OCA/sale-workflow/pull/1968#pullrequestreview-1087544016

simahawk replied:
solved?

jbaudoux commented:
What's the difference between this PR and https://github.com/OCA/sale-workflow/pull/1968 ?

mmequignon commented:
> What's the difference between this PR and #1968 ?

this refactoring commit, mostly https://github.com/OCA/sale-workflow/pull/1967/commits/88476b53a3a58af2beabe2e0012dff14c3cbda84
And the one moving "all" `sale.order.line` methods to `@api.model`, in order to be used from elsewhere.

mmequignon committed:
sale_delivery_date: fix picking report delivery date when late

jbaudoux commented on sale_delivery_date/models/sale_order_line.py:
```diff
@@ -5,176 +5,341 @@
+        # 1) commitment date is set, compute date_planned from date_deadline
+        # 2) commitment date isn't set, update date_deadline according to
+        #    cutoff / calendar / delivery window, and compute date_planned
+        if self.order_id.commitment_date:
```
As you now depend on sale_order_line_date
```suggestion
        if self.commitment_date or self.order_id.commitment_date:
```

mmequignon replied:
<s>Mah, we shouldn't need that, as order's commitment date is computed based on its line's commitment_dates.</s>

I was wrong this is not the case...
I think we should drop the dependency anyway.
@sebalix why did we added this `sale_order_line_date` dep? I'm not sure to understand why, according to this commit. Could we drop it?
https://github.com/OCA/sale-workflow/pull/1967/commits/7b68a2e7d69de05fc12e4d345b74ea25121fc55a

sebalix replied:
Because of this (copy/paste):

> **[FIX] sale_delivery_date: add dependency]**
> 
> Add `sale_order_line_date` as dependency because this addon is overriding the same methods (`_onchange_commitment_dat`e and `_prepare_procurement_values`) and we want to ensure than `sale_delivery_date` takes the precedence over `sale_order_line_date`.
> 
> Also, the other feature of `sale_order_line_date` which sync the commitment_date between SO and its lines is welcome in `sale_delivery_date`.

https://github.com/OCA/sale-workflow/pull/1968

If you think it is not needed anymore, feel free to remove it.

jbaudoux commented on sale_delivery_date/models/sale_order_line.py:
```diff
@@ -5,176 +5,341 @@
+        # stock_available_to_promise_release uses the date_planned as the
+        # preparation date before the picking is release, and as the expedition date
+        # after it has been released.
+        # /TODO Put this in the roadmap
```
TODO

jbaudoux commented on sale_delivery_date/models/sale_order_line.py:
```diff
@@ -5,176 +5,341 @@
+        # To avoid this, from the earliest delivery date, we need to retrieve
+        # the various delays applied above in order to find the latest
+        # possible expedition datetime and from it, the latest preparation datetime.
+        # N
```
```suggestion
        #
```

jbaudoux commented on sale_delivery_date/models/sale_order_line.py:
```diff
@@ -5,176 +5,341 @@
             )
-        return
+            latest_expedition_date = datetime.combine(
+                latest_expedition_datetime, time.max
```
but `time.max` is not in the warehouse timezone. When you call `_get_latest_work_end_from_date_range` below, this is converted to local timezone, and could end-up on a different date

jbaudoux commented on sale_delivery_date/models/sale_order_line.py:
```diff
@@ -184,57 +349,72 @@ def _prepare_procurement_values_cutoff_time(
+        return partner.next_delivery_window_start_datetime(from_date=delivery_date)
+
+    @api.model
+    def _get_latest_work_end_from_date_range(
+        self, earliest_work_end, latest_expedition_date, calendar=False
+    ):
+        """Returns the nearest date in a calendar's attendance within a date range"""
```
`_get_latest_work_end_from_date_range` `Returns the nearest date` `_get_closest_work_time`
latest / nearest / closest ?
The docstring should be improved as I found it hard to understand what's happenning
```suggestion
    def _get_latest_work_end_from_date_range(
        self, earliest_work_end, latest_expedition_date, calendar=False
    ):
        """Returns the last open date of the range according to the given calendar of open days"""
```

jbaudoux commented on sale_delivery_date/models/sale_order_line.py:
```diff
@@ -184,57 +349,72 @@ def _prepare_procurement_values_cutoff_time(
+        self, earliest_work_end, latest_expedition_date, calendar=False
+    ):
+        """Returns the nearest date in a calendar's attendance within a date range"""
+        if calendar:
```
A bit easier to read:
```suggestion
        if not calendar:
           return latest_expedition_date
        ...
```

jbaudoux commented on sale_delivery_date/models/sale_order_line.py:
```diff
@@ -184,57 +349,72 @@ def _prepare_procurement_values_cutoff_time(
-        return new_date_planned
+            if latest_date_end:
+                return latest_date_end.astimezone(UTC).replace(tzinfo=None)
+            return earliest_work_end  # TODO latest_expedition_date?
```
earliest make sense

jbaudoux reviewed (changes requested):
To be continued...

jbaudoux commented on sale_delivery_date/models/res_partner.py:
```diff
@@ -47,6 +51,7 @@ def next_delivery_window_start_datetime(self, from_date=None, timedelta_days=Non
         if timedelta_days is None:
             timedelta_days = 7
```
7 is not mutable. Why don't you put this in the signature of the method?

grindtildeath replied:
Actually it's there already.

jbaudoux commented on sale_delivery_date/models/res_partner.py:
```diff
@@ -47,6 +51,7 @@ def next_delivery_window_start_datetime(self, from_date=None, timedelta_days=Non
         if timedelta_days is None:
             timedelta_days = 7
```
7 is not mutable. Why don't you put this in the signature of the method?

jbaudoux commented on sale_delivery_date/models/res_partner.py:
```diff
@@ -47,6 +51,7 @@ def next_delivery_window_start_datetime(self, from_date=None, timedelta_days=Non
         if timedelta_days is None:
             timedelta_days = 7
         if self.delivery_time_preference == "workdays":
+            # TODO tz
```
I don't get why we compute all the possibilities while we just need the first correct one.
I would have dropped the call to `is_in_delivery_window` (because this is computing the first element you already compute in each method computing all possibilities in the range) and made this method cover all `delivery_time_preference` possibilities:
```
if self.delivery_time_preference == "anytime":
    return from_date
elif self.delivery_time_preference == "workdays":
    return self._get_first_workday(...)
elif self.delivery_time_preference == "time_windows":
    return self._get_first_timewindow(...)
else:
    raise UserError("Invalid delivery_time_preference")

jbaudoux commented on sale_delivery_date/models/sale_order_line.py:
```diff
@@ -184,57 +349,72 @@ def _prepare_procurement_values_cutoff_time(
+        """Postpone a delivery date according to customer's delivery preferences"""
+        if partner.delivery_time_preference == "anytime":
+            return delivery_date
+        return partner.next_delivery_window_start_datetime(from_date=delivery_date)
```
This should be renamed into `partner._get_first_date_in_delivery_time_preference(from_date=delivery_date)

And the 2 lines above for the anytime case should be moved inside that method

jbaudoux commented on sale_delivery_date/models/res_partner.py:
```diff
@@ -79,13 +97,24 @@ def get_next_windows_start_datetime(self, from_datetime, to_datetime):
+        from_datetime_tz = from_datetime_utc.astimezone(tz)
+        to_datetime_utc = utc.localize(to_datetime)
+        to_datetime_tz = to_datetime_utc.astimezone(tz)
+        for this_datetime in date_range(
+            from_datetime_tz, to_datetime_tz, timedelta(days=1)
+        ):
```
This method could be simplified as we just need to find the first day for which a time window is defined. 
We don't care on the time.

mmequignon committed:
sale_delivery_date: expected delivery - Always recompute

mmequignon committed:
sale_delivery_date: Additional cornercases tests

mmequignon committed:
sale_delivery_date: Only deliver on open days

mmequignon committed:
sale_delivery_date: Fix late expeditions detection

mmequignon committed:
fixup! sale_delivery_date: Fix late expeditions detection

mmequignon committed:
sale_delivery_date: Fix singleton exceptions

mmequignon committed:
fixup! fixup! sale_delivery_date: Fix late expeditions detection

mmequignon committed:
sale_delivery_date: use now if no date is set

mmequignon committed:
sale_delivery_date: sale report only display commitment_date if set

mmequignon committed:
sale_delivery_date: Allow to ignore customer delivery preferences

rousseldenis commented:
@mmequignon What's the status of this ?

jbaudoux commented:
> @mmequignon What's the status of this ?

@rousseldenis We are redoing it differently in v16

mmequignon commented:
> > @mmequignon What's the status of this ?
> 
> @rousseldenis We are redoing it differently in v16

@jbaudoux Not quite. Having another customer using release channels for this doesn't invalidate this module.

@rousseldenis This is currently used a lot by one of our customers, they did lots of tests before this version has been deployed in product. IMO, this is good to go. I have to find time to fix those tests.

rousseldenis commented:
> > > @mmequignon What's the status of this ?
> > 
> > 
> > @rousseldenis We are redoing it differently in v16
> 
> @jbaudoux Not quite. Having another customer using release channels for this doesn't invalidate this module.
> 
> @rousseldenis This is currently used a lot by one of our customers, they did lots of tests before this version has been deployed in product. IMO, this is good to go. I have to find time to fix those tests.

Should be great in order to merge this.

sebalix committed:
sale_delivery_date: add cache to improve performance

This avoids a lot of SQL requests (several hundreds on a SO with 20-30 lines),
and reduce the time needed to open orders.

rousseldenis commented:
@mmequignon

mmequignon committed:
sale_delivery_date: Consider line's commitment_date

jbaudoux commented:
Rewritten with a new modular approach on v16.0
https://github.com/OCA/wms/pull/1022
